File: ios/Classes/NavigineIconMapObject.swift
import Foundation
import Navigine

public class NavigineIconMapObject: NSObject {
    private var id: Int32
    private var mapObject: NCIconMapObject
    private let methodChannel: FlutterMethodChannel!
    private let pluginRegistrar: FlutterPluginRegistrar!

    public required init(
        id: Int32,
        mapObject: NCIconMapObject,
        registrar: FlutterPluginRegistrar
    ) {
        self.id = id
        self.mapObject = mapObject

        methodChannel = FlutterMethodChannel(
            name: "navigine_sdk/navigine_icon_map_object_\(id)",
            binaryMessenger: registrar.messenger()
        )

        self.pluginRegistrar = registrar

        super.init()

        weak var weakSelf = self
        self.methodChannel.setMethodCallHandler({ weakSelf?.handle($0, result: $1) })
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setPosition":
            result(setPosition(call))
        case "setPositionAnimated":
            result(setPositionAnimated(call))
        case "setImage":
            result(setImage(call))
        case "setSize":
            result(setSize(call))
        case "getType":
            result(getType(call))
        case "setVisible":
            result(setVisible(call))
        case "setInteractive":
            result(setInteractive(call))
        case "setStyle":
            result(setStyle(call))
        case "setData":
            setData(call)
            result(nil)
        case "getData":
            result(getData(call))
        case "setTitle":
            result(setTitle(call))
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    public func getMapObject() -> NCIconMapObject {
        return mapObject
    }

    public func setPosition(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let locationPoint = Utils.locationPointFromJson(params["locationPoint"] as! [String : Any])

        return mapObject.setPosition(locationPoint)
    }

    public func setPositionAnimated(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let locationPoint = Utils.locationPointFromJson(params["locationPoint"] as! [String : Any])
        let duration = (params["duration"] as! NSNumber)
        let type = params["type"] as! Int

        return mapObject.setPositionAnimated(locationPoint, duration: duration.floatValue, type: NCAnimationType.init(rawValue: type)!)
    }

    public func setImage(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let image = params["image"] as! [String : Any]

        return mapObject.setBitmap(getIconImage(image))
    }

    public func setSize(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let width = (params["width"] as! NSNumber)
        let height = (params["height"] as! NSNumber)

        return mapObject.setSize(width.floatValue,  height: height.floatValue)
    }

    public func getType(_ call: FlutterMethodCall) -> Int {
        return mapObject.getType().rawValue
    }

    public func setVisible(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let visible = (params["visible"] as! NSNumber)

        return mapObject.setVisible(visible.boolValue)
    }

    public func setInteractive(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let interactive = (params["interactive"] as! NSNumber)

        return mapObject.setInteractive(interactive.boolValue)
    }

    public func setStyle(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let style = (params["style"] as! String)

        return mapObject.setStyle(style)
    }

    public func setData(_ call: FlutterMethodCall) {
        let params = call.arguments as! [String: Any]
        let data = (params["data"] as! FlutterStandardTypedData).data

        return mapObject.setData(data)
    }

    public func getData(_ call: FlutterMethodCall) -> Data {
        return mapObject.getData()
    }

    public func setTitle(_ call: FlutterMethodCall) -> Bool {
        let params = call.arguments as! [String: Any]
        let title = (params["title"] as! String)

        return mapObject.setTitle(title)
    }

    private func getIconImage(_ image: [String: Any]) -> UIImage {
        let type = image["type"] as! String
        let defaultImage = UIImage()

        if (type == "fromAssetImage") {
          let assetName = self.pluginRegistrar.lookupKey(forAsset: image["assetName"] as! String)

          return UIImage(named: assetName) ?? defaultImage
        }

        if (type == "fromBytes") {
          let imageData = (image["rawImageData"] as! FlutterStandardTypedData).data

          return UIImage(data: imageData) ?? defaultImage
        }

        return defaultImage
      }
}

File: android/src/main/java/com/navigine/naviginesdk/NavigineIconMapObject.java
package com.navigine.naviginesdk;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import androidx.annotation.NonNull;

import com.navigine.idl.java.AnimationType;
import com.navigine.idl.java.IconMapObject;
import com.navigine.idl.java.LocationPoint;

import java.io.InputStream;
import java.util.Map;

import io.flutter.FlutterInjector;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class NavigineIconMapObject implements MethodChannel.MethodCallHandler {
  private final int id;
  private final Context context;
  private final IconMapObject mapObject;
  private final MethodChannel methodChannel;

  public NavigineIconMapObject (
    int id,
    Context context,
    IconMapObject mapObject,
    BinaryMessenger messenger
  ) {
    this.id = id;
    this.mapObject = mapObject;
    this.context = context;

    methodChannel = new MethodChannel(messenger, "navigine_sdk/navigine_icon_map_object_" + id);
    methodChannel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
    switch (call.method) {
      case "setPosition":
        result.success(setPosition(call));
        break;
      case "setPositionAnimated":
        result.success(setPositionAnimated(call));
        break;
      case "setImage":
        result.success(setImage(call));
        break;
      case "setSize":
        result.success(setSize(call));
        break;
      case "getType":
        result.success(getType(call));
        break;
      case "setVisible":
        result.success(setVisible(call));
        break;
      case "setInteractive":
        result.success(setInteractive(call));
        break;
      case "setStyle":
        result.success(setStyle(call));
        break;
      case "setData":
        setData(call);
        result.success(null);
        break;
      case "getData":
        result.success(getData(call));
        break;
      case "setTitle":
        result.success(setTitle(call));
        break;


      default:
        result.notImplemented();
        break;
    }
  }

  public IconMapObject getMapObject() {
    return mapObject;
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setPosition(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);
    return mapObject.setPosition(Utils.locationPointFromJson((Map<String, Object>) params.get("locationPoint")));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setPositionAnimated(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);
    LocationPoint locationPoint = Utils.locationPointFromJson((Map<String, Object>) params.get("locationPoint"));
    float duration = ((Double) params.get("duration")).floatValue();
    int type = ((Integer) params.get("type"));
    return mapObject.setPositionAnimated(locationPoint, duration, AnimationType.values()[type]);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setImage(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Map<String, Object> image = (Map<String, Object>) params.get("image");
    return mapObject.setBitmap(getIconImage(image));
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setSize(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Double width = (Double) params.get("width");
    Double height = (Double) params.get("height");
    return mapObject.setSize(width.floatValue(), height.floatValue());
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public int getType(MethodCall call) {
    return mapObject.getType().ordinal();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setVisible(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Boolean visible = (Boolean) params.get("visible");
    return mapObject.setVisible(visible);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setInteractive(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    Boolean interactive = (Boolean) params.get("interactive");
    return mapObject.setInteractive(interactive);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setStyle(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    String style = (String) params.get("style");
    return mapObject.setStyle(style);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void setData(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    byte[] data = (byte[]) params.get("style");
    mapObject.setData(data);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public byte[] getData(MethodCall call) {
    return mapObject.getData();
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public boolean setTitle(MethodCall call) {
    Map<String, Object> params = ((Map<String, Object>) call.arguments);

    String title = (String) params.get("title");
    return mapObject.setTitle(title);
  }

  @SuppressWarnings({"ConstantConditions"})
  private Bitmap getIconImage(Map<String, Object> image) {
    String type = (String) image.get("type");
    Bitmap defaultImage = Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888);

    if (type.equals("fromAssetImage")) {
      String assetName = FlutterInjector.instance().flutterLoader().getLookupKeyForAsset((String) image.get("assetName"));

      try (InputStream i = this.context.getAssets().open(assetName)) {
        return BitmapFactory.decodeStream(i);

      } catch (java.io.IOException e) {
        return defaultImage;
      }
    }

    if (type.equals("fromBytes")) {
      byte[] rawImageData = (byte[]) image.get("rawImageData");
      Bitmap bitmap = BitmapFactory.decodeByteArray(rawImageData, 0, rawImageData.length);

      if (bitmap != null) {
        return bitmap;
      }

      return defaultImage;
    }

    return defaultImage;
  }
}
